pkg/controller/machineset/delete_policy.go

```diff
@@ -17,57 +17,116 @@ limitations under the License.
+	}
+	return deletePriority(float64(mustDelete) * (1.0 - math.Exp(-d.Seconds()/secondsPerTenDays)))
+}
+
+func newestDeletePriority(machine *v1alpha1.Machine) deletePriority {
+	if machine.DeletionTimestamp != nil && !machine.DeletionTimestamp.IsZero() {
+		return mustDelete
+	}
+	if machine.ObjectMeta.Annotations != nil && machine.ObjectMeta.Annotations[DeleteNodeAnnotation] != "" {
+		return mustDelete
+	}
+	if machine.Status.ErrorReason != nil || machine.Status.ErrorMessage != nil {
+		return mustDelete
+	}
+	return mustDelete - oldestDeletePriority(machine)
+}
+
+func randomDeletePolicy(machine *v1alpha1.Machine) deletePriority {
 	if machine.DeletionTimestamp != nil && !machine.DeletionTimestamp.IsZero() {
 		return mustDelete
 	}
+	if machine.ObjectMeta.Annotations != nil && machine.ObjectMeta.Annotations[DeleteNodeAnnotation] != "" {
+		return betterDelete
+	}
 	if machine.Status.ErrorReason != nil || machine.Status.ErrorMessage != nil {
 		return betterDelete
 	}
 	return couldDelete
 }
 
-// TODO: Define machines deletion policies.
-// see: https://github.com/kubernetes/kube-deploy/issues/625
+type sortableMachines struct {
+	machines []*v1alpha1.Machine
+	priority deletePriorityFunc
+}
+
+func (m sortableMachines) Len() int      { return len(m.machines) }
+func (m sortableMachines) Swap(i, j int) { m.machines[i], m.machines[j] = m.machines[j], m.machines[i] }
+func (m sortableMachines) Less(i, j int) bool {
+	return m.priority(m.machines[j]) < m.priority(m.machines[i]) // high to low
+}
+
 func getMachinesToDeletePrioritized(filteredMachines []*v1alpha1.Machine, diff int, fun deletePriorityFunc) []*v1alpha1.Machine {
 	if diff >= len(filteredMachines) {
 		return filteredMachines
 	} else if diff <= 0 {
 		return []*v1alpha1.Machine{}
 	}
 
-	machines := make(map[deletePriority][]*v1alpha1.Machine)
-
-	for _, machine := range filteredMachines {
-		priority := fun(machine)
-		machines[priority] = append(machines[priority], machine)
+	sortable := sortableMachines{
+		machines: filteredMachines,
+		priority: fun,
 	}
+	sort.Sort(sortable)
 
-	result := []*v1alpha1.Machine{}
-	for _, priority := range []deletePriority{
-		mustDelete,
-		betterDelete,
-		couldDelete,
-	} {
-		result = append(result, machines[priority]...)
-		if len(result) >= diff {
-			break
-		}
-	}
+	return sortable.machines[:diff]
+}
 
-	return result[:diff]
+func getDeletePriorityFunc(ms *v1alpha1.MachineSet) (deletePriorityFunc, error) {
+	// Map the Spec.DeletePolicy value to the appropriate delete priority function
+	switch msdp := v1alpha1.MachineSetDeletePolicy(ms.Spec.DeletePolicy); msdp {
+	case v1alpha1.RandomMachineSetDeletePolicy:
+		return randomDeletePolicy, nil
+	case v1alpha1.NewestMachineSetDeletePolicy:
+		return newestDeletePriority, nil
+	case v1alpha1.OldestMachineSetDeletePolicy:
+		return oldestDeletePriority, nil
+	default:
+		return nil, errors.Errorf("Unsupported delete policy %s. Must be one of 'Random', 'Newest', or 'Oldest'", msdp)
```

Comment:
late comment, but we have to default to a policy if the value is not set, i.e.,
```
	case "":
		return randomDeletePolicy, nil
```

Reply:
@ntfrnzn (not sure if this change went in after your comment, but) if you look a couple of lines up, there is a default for unset/`""`. The `default` case is for when the user specifies an invalid policy name such as `abc123`.